### PR TITLE
[No QA] Create Incremental 'withOnyx' migration CI check

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details
As part of our ongoing transition strategy from `withOnyx` to `useOnyx` (discussed in this [proposal](https://expensify.slack.com/archives/C01GTK53T8Q/p1725905735105989)), we are introducing a new CI check.

When a contributor modifies a file that contains `withOnyx` usages, this check will be triggered. the CI will fail on your PR, regardless of whether you introduced that usage.

To make things easier for contributors, you are not expected to migrate all `withOnyx` HOCs across every file modified in a PR. Your objective is to simply decrease the overall usage of withOnyx. This way, you could focus on migrating just one file per PR, which gradually contributes to our overall goal.

### Fixed Issues
$ https://github.com/Expensify/App/issues/49102
PROPOSAL: N/A

### Tests
- [Create a file without withOnyx](https://github.com/Expensify/App/pull/48885/commits/b98527d94d69f8939daa1b39109e3b0adad54f1e) - [check is successful](https://github.com/Expensify/App/actions/runs/10792169388/job/29931252191)
- [Create a file with withOnyx](https://github.com/Expensify/App/pull/48885/commits/2ffd307c91561e233267d55e304f9b8a78ce94e4) - [check failed](https://github.com/Expensify/App/actions/runs/10792178069/job/29931281479)
- [Edit existing file without migrating to useOnyx](https://github.com/Expensify/App/pull/48885/commits/7643c7d3b767f00c5a9983402c91e22d805d905e) - [check failed](https://github.com/Expensify/App/actions/runs/10792189723/job/29931316644)
- [Edit existing file and migrate to useOnyx](https://github.com/Expensify/App/pull/48885/commits/6f2646e4a8b45369d69a234500ac257440e5d8ef) - [check is successful](https://github.com/Expensify/App/actions/runs/10792194728/job/29931331422)
- [No modifications, should not even trigger the check](https://github.com/Expensify/App/pull/48885/commits/7643c7d3b767f00c5a9983402c91e22d805d905e) - check wasn't triggered
- [Multiple files with one migration](https://github.com/Expensify/App/pull/48885/commits/19960bdc22a49495254bf2e23f59f99c353273b2) - [check successful](https://github.com/Expensify/App/actions/runs/10795112910/job/29940988105?pr=48885)
- [Multiple files modified without migrating](https://github.com/Expensify/App/pull/48885/commits/f69c458ac28fc3e53a0b4bd54c32faa18bb0173b) - [Check failed](https://github.com/Expensify/App/actions/runs/10795168444/job/29941190343)

### Offline tests
N/A

### QA Steps
N/A

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
### Screenshots/Videos

N/A

### CODE REVIEW

**Comment by fabioh8010** [comment]
@blazejkustra Could you put an example of editing multiple files with `withOnyx` so we can see how is the output?

**Review by fabioh8010 — APPROVED** [review]
🚀

File: .github/scripts/withOnyxMigration.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+# Set IFS to newline to properly handle file names with spaces.
+# This makes 'for' loop treat each line as a separate item.
+IFS=$'\n'
+
+# Fetch the main branch
+git fetch origin main --no-tags --depth=1
+
+# Get the list of modified and added files
+modified_files=$(git diff --name-only --diff-filter=AM origin/main HEAD -- '*.ts' '*.tsx')
```
Comment (Kicu) [inline .github/scripts/withOnyxMigration.sh:16]: actually I'm wondering if we need to check `ts` files at all, I think we can check only tsx

File: .github/workflows/withOnyxMigration.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+name: Incremental 'withOnyx' migration
+
+on:
+  workflow_call:
+  pull_request:
+    types: [opened, synchronize]
+    branches-ignore: [staging, production]
+    paths: ['**.ts', '**.tsx']
```
Comment (Kicu) [inline .github/workflows/withOnyxMigration.yml:8]: same here, check only for `**.tsx`

File: .github/scripts/withOnyxMigration.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+# Count instances of 'withOnyx' on main branch and current branch
+count_main=0
+count_current=0
+
+echo -e "\nWe're gradually moving from 'withOnyx' to 'useOnyx'. 'useOnyx' hook simplifies the code, offers performance benefits and improves TypeScript compilation time."
+echo -e "For more details, see https://expensify.slack.com/archives/C01GTK53T8Q/p1725905735105989.\n"
+echo -e "${BOLD_WHITE}If you are blocked by this on something that is very urgent, you can always ignore this check and merge with it failing.\n"
```
Comment (Kicu) [inline .github/scripts/withOnyxMigration.sh:24]:
```suggestion
echo -e "${BOLD_WHITE}If this check is blocking you on something that is very urgent, you can always ignore it.\n"
```

maybe just a simpler and shorter sentence is better?

**Review by Kicu — APPROVED** [review]
LGTM 👍 only minor comments from me
I don't have strong opinions on writing shell scripts, so for me if this works then it's good.
Counting via grep seems straightforward.

**Comment by roryabraham** [comment]
Think I'm going to close this in favor of https://github.com/Expensify/App/pull/49125, which is a bit simpler and more general-purpose

**Pull request closed without merging**